### PR TITLE
[android] Fix intermittent GC of offline callbacks

### LINKED ISSUE #14297: ResourcesCachePathChangeCallback is sporadically not returning
Sometimes the `ResourcesCachePathChangeCallback` is not returning. After some debugging I found out, that the `FileUtils.OnCheckFileWritePermissionListener` is gone (null) in the `FileUtils.CheckFileWritePermissionTask`. Looks like a strange JNI behavior. 

The problem occurs more often when the resources cache path should be changed directly after the initialization of Mapbox and also on older devices.

### Steps to reproduce

 1. Initialize Mapbox
 2. Set custom resource cache path

### Expected behavior

The cache path should be set every time.

### Actual behavior

Sometimes the cache path isn't set and the callback doesn't return.

### Configuration

**Android versions:** all
**Device models:** all
**Mapbox SDK versions:** 7.3.0

### PR DESCRIPTION
**Pull request by Guardiola31337**
- Fixes #14297 

We took the second approach explained in https://github.com/mapbox/mapbox-gl-native/issues/14297#issuecomment-488021326

> - Remove the `WeakReference`s and reset the callbacks when `onCancelled` and `onPostExecute` so they're properly GC'ed

Also just to make sure that an exception thrown inside the callback doesn't prevent us from cleaning the listener, we're making the reference local and clearing the field one before calling it. Extra safe 😅 

We've tested this implementation downstream in https://github.com/mapbox/mapbox-navigation-android/pull/1895 and no leaks were reported by LeakCanary.

cc @zugaldia

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java [hunk 1]
```diff
@@ -225,7 +225,7 @@ public void run() {
    */
   public void mergeOfflineRegions(@NonNull String path, @NonNull final MergeOfflineRegionsCallback callback) {
     final File src = new File(path);
-    new FileUtils.CheckFileReadPermissionTask(new FileUtils.OnCheckFileReadPermissionListener() {
+    final FileUtils.OnCheckFileReadPermissionListener fileReadPermissionListener = new FileUtils.OnCheckFileReadPermissionListener() {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java]: We can revert this change back if needed (that was from previous testing / debugging)

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java [hunk 2]
```diff
@@ -250,18 +250,19 @@ public void onError() {
         // path not readable, abort
         callback.onError("Secondary database needs to be located in a readable path.");
       }
-    }).execute(src);
+    };
+    new FileUtils.CheckFileReadPermissionTask(fileReadPermissionListener).execute(src);
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java]: Same here ☝️

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java [hunk 3]
```diff
@@ -225,7 +225,8 @@ public void run() {
    */
   public void mergeOfflineRegions(@NonNull String path, @NonNull final MergeOfflineRegionsCallback callback) {
     final File src = new File(path);
-    new FileUtils.CheckFileReadPermissionTask(new FileUtils.OnCheckFileReadPermissionListener() {
+    final FileUtils.OnCheckFileReadPermissionListener fileReadPermissionListener =
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineManager.java]: We can revert this change back if needed (that was from previous testing / debugging)

**Comment by LukasPaczos** [comment]
Thanks for looking into this @Guardiola31337 @danesfeder!

I also took his branch for a spin and this approach seems to still be running into leaks. Let's assume this scenario:
```
@Override
protected void onCreate(Bundle savedInstanceState) {
  super.onCreate(savedInstanceState);
  setContentView(R.layout.activity_feature_overview);

  new FileUtils.CheckFileWritePermissionTask(new FileUtils.OnCheckFileWritePermissionListener() {
    @Override
    public void onWritePermissionGranted() {
     Logger.e("mbgl", "granted");
    }

    @Override
    public void onError() {
      Logger.e("mbgl", "error");
    }
  }).execute(new File(""));

  finish();
}
``` 

Because in this PR we are holding a hard reference to the callback, we're going to inevitably leak the context (or any caller). Adding enough `Thread.sleep` to the async task makes this issue easily reproducible and is being picked up by the leak canary.

Currently, on `master` the same example would result in the callback being GC'ed. By treating all callback references as weak, we are ensuring to not leak and giving users the option to store the callback locally and make sure it's delivered:
```
private final FileUtils.OnCheckFileWritePermissionListener listener = new FileUtils.OnCheckFileWritePermissionListener() {
  @Override
  public void onWritePermissionGranted() {
    Logger.e("mbgl", "granted");
  }

  @Override
  public void onError() {
    Logger.e("mbgl", "error");
  }
};

@Override
protected void onCreate(Bundle savedInstanceState) {
  super.onCreate(savedInstanceState);
  setContentView(R.layout.activity_feature_overview);
  new FileUtils.CheckFileWritePermissionTask(listener).execute(new File(""));
  finish();
}
```
This approach is going to leak as well, so the user needs to clear the reference:
```
@Override
protected void onDestroy() {
  super.onDestroy();
  listener = null;
}
```

The combination of the second and third code blocks seems to me like the only way to cleanly handle the callback. The `register`/`unregister` methods combo would work in the same manner but it doesn't fit the exposed interface.
All the above thoughts are true for **any callback used in the offline merge/resource path change flow**.

The offline merge flow is built out of 3 paths:
- path writeable -> merge
- path not writeable -> path readable -> copy to a writeable path -> merge
- path not writeable -> path not readable -> fail

Each of the steps contains a separate callback and since the classes are not lifecycle-aware we can either lose a reference or leak on any of those steps. Similarly for the resources path change flow.

### Conclusion
In order to make the process leak-proof and harden the callback references, I'm proposing to streamline the process to one worker thread and simplify kept references to only the weak references of `Context` and user's final `Callback`. This makes sure that the process is going to be valid for as long as the user holds a reference to the `Callback`.

An alternative would be holding a hard reference to the `Callback` which in turn would force a user to wrap it and pass a weak reference to this wrapper as we are doing with the location APIs:
https://github.com/mapbox/mapbox-gl-native/blob/e69efb996d32c86f00b08878a61c6f0a47c81932/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L1425-L1438

Let me know if you agree with the causes and the proposed solutions @Guardiola31337 @danesfeder @tobrun. If so, do you prefer any approach over another? Either way, we should document the actions that need to be done when using mentioned APIs.

**Comment by Guardiola31337** [comment]
Yeah, you're right @LukasPaczos 

The problem here is that `Thread`s and `AsyncTask`s do not support cancellation for blocking operations like the ones we’re doing, which means that without using something more sophisticated like Rx or adding an `onDestroy` setter that makes the references `null` wisely https://github.com/mapbox/mapbox-gl-native/pull/14515#issuecomment-487085055 (with the caveat that you already pointed out `The register/unregister methods combo would work in the same manner but it doesn't fit the exposed interface.`), we will not be able to release references before the operation finishes. These operations take little time to finish (except for `copyTempDatabaseFile`), so it shouldn’t be noticeable to users. It’s not great, but it’s acceptable, and better than not receiving a callback or leaking forever.

Another approach here would be trying to convert those APIs into non-static but maybe that's a bigger refactor and it's out of scope at this point. In any case, have we considered fixing https://github.com/mapbox/mapbox-gl-native/issues/14297 and https://github.com/mapbox/mapbox-gl-native/issues/14485 creating new non-static internal APIs and use them from `OfflineManager` - although we'll be duplicating behavior I think it's a good tradeoff (for now).

That being said, I'm down for a bigger refactor too. Could you submit a PR @LukasPaczos so I can take a look of what exactly you have in mind?

**Comment by LukasPaczos** [comment]
> Could you submit a PR @LukasPaczos so I can take a look of what exactly you have in mind?

Sure @Guardiola31337, let me follow up with a PR draft and we can discuss there.

**Comment by LukasPaczos** [comment]
Closing in favor of https://github.com/mapbox/mapbox-gl-native/pull/14601. Thanks a lot for starting the work and discussion here @Guardiola31337!

https://github.com/mapbox/mapbox-gl-native/pull/14601 is going o go out next Wednesday with `7.5.0-beta.1` and I'd be great if you could test it downstream (or the `7.5.0-SNAPSHOT` that's already available).
